lib/scholar/linear/linear_regression.ex

```diff
@@ -0,0 +1,94 @@
+defmodule Scholar.Linear.LinearRegression do
+  @moduledoc """
+  Ordinary least squares linear regression.
+  """
+  import Nx.Defn
+
+  @derive {Nx.Container, containers: [:coefficients]}
+  defstruct [:coefficients]
+
+  opts = [
+    sample_weights: [
+      type: {:list, {:custom, Scholar.Options, :positive_number, []}}
+    ]
+  ]
+
+  @opts_schema NimbleOptions.new!(opts)
+
+  @doc """
+  Fits a linear regression model for sample inputs `a` and
+  sample targets `b`.
+  """
+  deftransform fit(a, b, opts \\ []) do
+    opts = NimbleOptions.validate!(opts, @opts_schema)
+
+    opts =
+      [
+        sample_weights_flag: opts[:sample_weights] != nil
+      ] ++
+        opts
+
+    {sample_weights, opts} = Keyword.pop(opts, :sample_weights, 1.0)
+    sample_weights = Nx.tensor(sample_weights)
+
+    fit_n(a, b, sample_weights, opts)
+  end
+
+  defnp fit_n(a, b, sample_weights, opts \\ []) do
+    {a, b} = preprocess_data(a, b, sample_weights, opts)
+
+    {a, b} =
+      if opts[:sample_weights_flag] do
+        rescale(a, b, sample_weights)
+      else
+        {a, b}
+      end
+
+    coeff = lstsq(a, b)
+    %__MODULE__{coefficients: Nx.transpose(coeff)}
+  end
+
+  @doc """
+  Makes predictions with the given model on inputs `x`.
+  """
+  defn predict(%__MODULE__{coefficients: coeff}, x) do
+    Nx.dot(x, coeff)
+  end
+
+  # Implements sample weighting by rescaling inputs and
+  # targets by sqrt(sample_weight).
+  defnp rescale(x, y, sample_weights) do
+    n = Nx.axis_size(x, 0)
+
+    sample_weights =
+      case Nx.shape(sample_weights) do
+        {} ->
+          Nx.broadcast(sample_weights, {n})
+
+        _ ->
+          sample_weights
+      end
+
+    scale = Nx.sqrt(sample_weights) |> Nx.make_diagonal()
+
+    {Nx.dot(scale, x), Nx.dot(scale, y)}
+  end
+
+  # Implements ordinary least-squares by estimating the
+  # solution A to the equation A.X = b.
+  defnp lstsq(a, b) do
+    pinv = Nx.LinAlg.pinv(a)
+    Nx.dot(pinv, b)
```

Comment:
I think you can avoid the transpose on L46 if you do:
```suggestion
    Nx.dot(b, [0], pinv, [1])
```

This doesn't work for a batched version (you'd need to pass batch axes as well), but I don't think you're supporting batches here anyway.

```elixir
t1 = Nx.random_uniform({5, 4})

t2 = Nx.random_uniform({4, 6})

t3 = Nx.dot(t1, t2)
t4 = Nx.dot(t2, [-2], t1, [-1])

Nx.transpose(t3) == t4 # this returns true (using binary backend)
```

Reply:
It is ok to not support batching, let's just make sure to raise on the shapes.

Reply:
Yes, I didn't mean that batching should have to be supported here